### DISCUSSION TITLE
First-class switch statement

### DISCUSSION THREAD

**Discussion by rachitnigam** [cucapra/futil#630] [Ideas]
Calyx currently supports multi-armed `if` statements but often times, frontend code would benefit from a first class switch statement instead. For example, this pattern seems to come up a few times:
```
par {
  if x.out with cond0 { .. }
  if x.out with cond1 { .. }
  ...
}
```

Where `cond0` and `cond1` are really just doing the same comparison (think unrolled accesses generated from the Dahlia frontend). A better solution would be to have a switch-case style construct:
```
switch x.out {
  0 => { ... }
  1 => { ... }
}
```

`if` can remain surface syntax that is compiled to `switch` statements.

**Comment by sampsyo** [comment]
Seems pretty reasonable! But one thing I’m a little confused about: won’t we need some (moral equivalent of) the `with` clause? That is, the construct will need to know which group to activate to obtain `x.out`, right? (Even if it only activates it once and shares the result.)

**Comment by rachitnigam** [comment]
Sure, we can have a `switch <port> with <group>` construct when the `<group>` is a combinational group. I almost titled this discussion "Calyx 2" because we have now proposed a bunch of substantial changes to the language and the semantics that would be worth incorporating.
